### PR TITLE
EDM-226: Agent Update/Rollback e2e tests

### PR DESCRIPTION
**Pull request by kkyrazis**
E2E Tests for the remaining test cases defined under the EDM-226 block in polarion.

1. I am proposing splitting the e2e's harness.go file into multiple files. It's already fairly lengthy and likely to become even larger over time. I've defined a harness_fleet.go which just moves separates out the fleet specific functions. The end state could be something like:
* `harness.go` => shared functionality
* `harness_[resource].go` => resource specific functions.
If you don't like it just let me know and I'll move it back.

2. For whatever reason .gitignore was not working for the e2e base image. Everytime I built git would show that I had changes. I've deleted the template file and updated the base image to require a REGISTRY_ADDRESS arg that the create_agent_images.sh script now supplies instead of echoing the config into the image file directly. Again, if you don't like it just let me know and I'll move it back and deal with the gitignore issues.

3. The rollback e2e test works around a known bug that is being worked on in EDM-1365. An assertion that should be added when that bug is fixed is currently commented out.

## Summary by CodeRabbit

- **New Features**
  - Added new end-to-end tests for device and fleet conflict detection, update handling, and rollback scenarios.
  - Introduced new test harness utilities for repository management, device labeling, log retrieval, and generic resource waiting.
  - Added support for building and testing with a deliberately broken device image to verify failure handling.
- **Bug Fixes**
  - Improved test coverage for device update failures and rollback attempts, including log verification.
- **Refactor**
  - Restructured test harness code for better reusability and modularity.
  - Moved fleet-related test helpers to a dedicated file and enhanced error handling.
- **Chores**
  - Updated container build scripts and Dockerfiles to support dynamic registry configuration and new test images.
  - Adjusted `.gitignore` to better handle certificate and key files in test directories.
  - Removed obsolete container build templates.

### CODE REVIEW

File: test/e2e/agent/agent_test.go [hunk 1]
```diff
@@ -71,6 +71,109 @@ var _ = Describe("VM Agent behavior", func() {
 			Eventually(harness.GetDeviceWithStatusSystem, TIMEOUT, POLLING).WithArguments(
 				enrollmentID).ShouldNot(BeNil())
 		})
+		It("Should report a message when a device is assigned to multiple fleets", Label("75992", "sanity"), func() {
+			const (
+				fleet1Name  = "fleet1"
```
Comment (eldar101) [inline test/e2e/agent/agent_test.go:76]: we have templates for fleets...see other tests on how we create fleets from our examples folder

File: test/e2e/agent/agent_test.go [hunk 2]
```diff
@@ -71,6 +71,109 @@ var _ = Describe("VM Agent behavior", func() {
...
+				fleet2Name  = "fleet2"
+				fleet1Label = "region"
+				fleet2Label = "environment"
+				fleet1Value = "world"
+				fleet2Value = "prod"
+				timeout     = "2m"
+			)
+			deviceId, _ := harness.EnrollAndWaitForOnlineStatus()
+			err := harness.WaitForDeviceNewRenderedVersion(deviceId, 1)
```
Comment (siserafin) [inline test/e2e/agent/agent_test.go]: You can store the first device version in a variable. Then, for the next versions, we have functions like PrepareNextDeviceVersion, to get the next version.

File: test/e2e/agent/agent_test.go (hunk 1 above)
Comment (siserafin) [reply test/e2e/agent/agent_test.go:76]: @eldar101  example fleets are useful for testing features where the specific device configuration is not important (e.g client or filtering tests). However, in other features, we create a fleet with custom device specifications, which are applied to devices matching the label selector. In this test we are testing some edge situations, for example the behavior of the system in case a device has labels matching multiple fleets, or fleets with the same label selector.

File: test/e2e/agent/agent_test.go [hunk 3]
```diff
@@ -71,6 +71,109 @@ var _ = Describe("VM Agent behavior", func() {
...
+			}, timeout)
+
+			// verify that the conflicting fleets are applied to the error message
+			device, err := harness.GetDevice(deviceId)
+			Expect(err).ToNot(HaveOccurred())
+			cond := v1alpha1.FindStatusCondition(device.Status.Conditions, v1alpha1.DeviceMultipleOwners)
+			Expect(cond).ToNot(BeNil())
+			Expect(cond.Message).Should(And(ContainSubstring(fleet1Name), ContainSubstring(fleet2Name)))
+			// No updates from the fleet should have been applied
+			Expect(device.Status.Config.RenderedVersion).To(Equal("1"))
```
Comment (siserafin) [inline test/e2e/agent/agent_test.go]: here you can use the variable created before

File: test/e2e/agent/agent_test.go [hunk 4]
```diff
@@ -71,6 +71,109 @@ var _ = Describe("VM Agent behavior", func() {
...
+			harness.SetLabelsForDevice(deviceId, nil)
+			harness.WaitForDeviceContents(deviceId, "multiple owners condition should be removed", func(device *v1alpha1.Device) bool {
+				return e2e.ConditionStatusExists(device.Status.Conditions, v1alpha1.DeviceMultipleOwners, v1alpha1.ConditionStatusFalse)
+			}, timeout)
+
+			By("adding a label to a matching fleet, the device should update its rendered version")
+			harness.SetLabelsForDevice(deviceId, map[string]string{
+				fleet1Label: fleet1Value,
+			})
+			err = harness.WaitForDeviceNewRenderedVersion(deviceId, 2)
```
Comment (siserafin) [inline test/e2e/agent/agent_test.go]: Use a variable for the version

File: test/e2e/agent/agent_updates_test.go [hunk 5]
```diff
@@ -157,5 +161,139 @@ var _ = Describe("VM Agent behavior during updates", func() {
...
+						*device.Spec.Config = append(*device.Spec.Config, configProviderSpec)
+					}
+					logrus.Infof("Updating %s with config %s", deviceId, *device.Spec.Config)
+				})
+				expectedVersion := currentVersion + i + 1
+				desc := fmt.Sprintf("Updating to desired renderedVersion: %d", expectedVersion)
+				By(fmt.Sprintf("Waiting for update %d to be picked up", i+1))
+				harness.WaitForDeviceContents(deviceId, desc,
+					func(device *v1alpha1.Device) bool {
+						version, err := strconv.Atoi(device.Status.Config.RenderedVersion)
```
Comment (siserafin) [inline test/e2e/agent/agent_updates_test.go]: Maybe here you could use  harness.GetCurrentDeviceRenderedVersion

File: test/e2e/agent/agent_updates_test.go [hunk 6]
```diff
@@ -157,5 +161,139 @@ var _ = Describe("VM Agent behavior during updates", func() {
...
+						}
+						// The update has already applied
+						if version == expectedVersion {
+							return true
+						}
+						cond := v1alpha1.FindStatusCondition(device.Status.Conditions, v1alpha1.DeviceUpdating)
+						if cond == nil {
+							return false
+						}
+						// send another update if we're in this state
```
Comment (siserafin) [inline test/e2e/agent/agent_updates_test.go]: Nice!

File: test/e2e/agent/agent_updates_test.go [hunk 7]
```diff
@@ -157,5 +161,139 @@ var _ = Describe("VM Agent behavior during updates", func() {
 
 			logrus.Info("Went back to base image and checked that there is no application now👌")
 		})
+		It("Should resolve to the latest version when multiple updates are applied", Label("77672", "sanity"), func() {
```
Comment (siserafin) [inline test/e2e/agent/agent_updates_test.go:164]: Nice test! Do you think it would be possible to use some local helper functions to make the steps more readable in the test body?

File: test/scripts/agent-images/create_agent_images.sh [hunk 8]
```diff
@@ -52,7 +46,13 @@ build_images() {
 
         echo -e "\033[32mCreating image ${FINAL_REF} with BUILD_TYPE=${BUILD_TYPE} \033[m"
 
-        podman build ${BUILD_ARGS:+${BUILD_ARGS}} -f "${containerfile_path}" -t "${container_name}" .
+        # apply image specific args here
```
Comment (siserafin) [inline test/scripts/agent-images/create_agent_images.sh:49]: LGTM

File: test/e2e/agent/agent_test.go (hunk 1 above)
Comment (kkyrazis) [reply test/e2e/agent/agent_test.go:76]: What Silvia said. It felt unnecessary to create/use templates for this specific test as the contents of the fleet are largely unimportant and I'm just checking a few status edge cases.

File: test/e2e/agent/agent_updates_test.go (hunk 7 above)
Comment (kkyrazis) [reply test/e2e/agent/agent_updates_test.go:164]: Sure, I'll spend some time cleaning this test up.

File: test/harness/e2e/harness_fleet.go [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+			},
+		}
+
+	default:
+		return fmt.Errorf("first parameter must be either FleetSpec or LabelSelector")
+	}
+
+	_, err := h.Client.ReplaceFleetWithResponse(h.Context, testFleetName, testFleet)
+	return err
+}
```
[the remark this answers is not in the document]
Comment (kkyrazis) [reply test/harness/e2e/harness_fleet.go:91]: No, the PUT /fleets/{name} endpoint can create fleets if one does not exist and will return a 201. It behaves as an upsert.

File: test/e2e/agent/agent_updates_test.go (hunk 5 above)
Comment (kkyrazis) [reply test/e2e/agent/agent_updates_test.go]: `GetCurrentDeviceRenderedVersion` makes another http request to refetch the device. At this point we already have the fresh device. I'll add a utility function to do the conversion though.

File: test/harness/e2e/harness.go [hunk 10]
```diff
@@ -1238,25 +1178,25 @@ func (h *Harness) FixNetworkFailure() error {
...
 
 // CheckRunningContainers verifies the expected number of running containers on the VM.
-func (h Harness) CheckRunningContainers() (string, error) {
+func (h *Harness) CheckRunningContainers() (string, error) {
 	out, err := h.VM.RunSSH([]string{"sudo", "podman", "ps", "|", "grep", "Up", "|", "wc", "-l"}, nil)
 	if err != nil {
 		return "", err
 	}
 	return out.String(), nil
 }
```
[the remark this answers is not in the document]
Comment (kkyrazis) [reply test/harness/e2e/harness.go:1350]: That is outside the scope of this PR

File: test/e2e/agent/agent_updates_test.go [hunk 11]
```diff
@@ -157,5 +161,153 @@
...
+			err = harness.ReplaceRepository(spec, repoMetadata)
+			Expect(err).NotTo(HaveOccurred())
+			// clean up the repository when we're done with it
+			defer func() {
+				err := harness.DeleteRepository(*repoMetadata.Name)
+				if err != nil {
+					logrus.Errorf("Failed to delete repository %s: %v", *repoMetadata.Name, err)
+				}
+			}()
+
```
[the remark this answers is not in the document]
Comment (kkyrazis) [reply test/e2e/agent/agent_updates_test.go:178]: These are defined in another file in this package

File: test/e2e/agent/agent_updates_test.go [hunk 12]
```diff
@@ -157,5 +161,153 @@ var _ = Describe("VM Agent behavior during updates", func() {
...
+					var configProviderSpec v1alpha1.ConfigProviderSpec
+					err := factory(&configProviderSpec)
+					Expect(err).ToNot(HaveOccurred())
+					// append, not overwrite, new specs
+					specs := lo.FromPtr(device.Spec.Config)
+					specs = append(specs, configProviderSpec)
+					device.Spec.Config = &specs
+					logrus.Infof("Updating %s with config %+v", deviceId, *device.Spec.Config)
+				})
+				expectedVersion := currentVersion + 1
```
[the remark this answers is not in the document]
Comment (kkyrazis) [reply test/e2e/agent/agent_updates_test.go:207]: lo.FromPtr handles this for me.

File: test/harness/e2e/harness_fleet.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (kkyrazis) [reply test/harness/e2e/harness_fleet.go:91]: While not a terrible suggestion, I'll leave the behavior as it is currently implemented until such a time that handling conflicts is necessary.

**Review by Samudelacruz77 — APPROVED** [review]
LGTM
